### PR TITLE
UI for "ParaTime wizard"

### PR DESCRIPTION
**Pull request by buberdds**
Next PRs will cover:

1) 
- deposit to/from Cipher/Emerald/Sapphire ParaTime
- withdraw EMVc withdraws need an extra step with recipient address.

2) 
- allow access route only when wallet is 'open' 
- refresh account after successful action. We probably don't want to use current loader (with full page overflow)
- add collapsible advance section with fees
- transaction failure message / make success/error status more prominent in the last step
- add more warnings (old extension modals) and additional ledger info
- fix ETH value with comma not parsed correctly
- CR:I think this should be BigInt(amount) > BigInt(value), and no need for EVM specific variant
- unit tests

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -57,3 +57,79 @@ export const config: BackendConfig = {
...
+}
+
+type ParaTimeConfig = {
+  [key in ParaTimes]: ParaTimeLayer
+}
+
+export const paraTimesConfig: ParaTimeConfig = {
+  [ParaTimes.Cipher]: { ...cipherConfig },
+  [ParaTimes.Emerald]: { ...emeraldConfig },
+  [ParaTimes.Sapphire]: { ...sapphireConfig },
```
Comment (lukaw3d) [inline src/config.ts:134]: Needless object spread?

File: src/app/pages/ParaTimesPage/useParaTimesNavigation.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  const navigateToDeposit = useCallback(() => dispatch(paraTimesActions.navigateToDeposit()), [dispatch])
+  const navigateToWithdraw = useCallback(() => dispatch(paraTimesActions.navigateToWithdraw()), [dispatch])
+  const navigateToParaTimes = useCallback(() => dispatch(paraTimesActions.navigateToParaTimes()), [dispatch])
+  const navigateToRecipient = useCallback(() => dispatch(paraTimesActions.navigateToRecipient()), [dispatch])
+  const navigateToAmount = useCallback(() => dispatch(paraTimesActions.navigateToAmount()), [dispatch])
+  const navigateToConfirmation = useCallback(
+    () => dispatch(paraTimesActions.navigateToConfirmation()),
+    [dispatch],
+  )
+  const navigateToSummary = useCallback(() => dispatch(paraTimesActions.navigateToSummary()), [dispatch])
```
[the remark this answers is not in the document]
Comment (buberdds) [reply src/app/pages/ParaTimesPage/useParaTimesNavigation.ts:17]: I am not sure about this part of redux. If useDispatch is stable just like useState then yes we can remove useCalback hooks.

File: src/app/components/AmountFormatter/index.tsx [hunk 3]
```diff
@@ -9,10 +9,12 @@ import { memo } from 'react'
 import { useSelector } from 'react-redux'
 import { Text } from 'grommet'
 import { StringifiedBigInt } from 'types/StringifiedBigInt'
-import { formatBaseUnitsAsRose } from 'app/lib/helpers'
+import { formatBaseUnitsAsRose, formatGweiAsWrose } from 'app/lib/helpers'
+import { grommetCustomTheme } from '../../../styles/theme/ThemeProvider'
 
 export interface AmountFormatterProps {
   amount: StringifiedBigInt | null
+  baseUnit?: boolean
```
[the remark this answers is not in the document]
Comment (buberdds) [reply src/app/components/AmountFormatter/index.tsx:16]: What prop name do you prefer?

File: src/app/pages/ParaTimesPage/ParaTimeFormFooter/index.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            onClick={secondaryAction}
+            plain
+            style={{ textAlign: 'center', fontSize: '14px', textDecoration: 'underline', color: 'brand' }}
+          />
+        )}
+      </Box>
+
+      {withNotice && (
+        <Text size="small">
+          {t('paraTimes.footer.notice', '* EVMc - Ethereum Virtual Machine compatible')}
```
[the remark this answers is not in the document]
Comment (buberdds) [reply src/app/pages/ParaTimesPage/ParaTimeFormFooter/index.tsx:52]: We will update messaging when we will  have an established abbreviation

File: src/app/components/AmountFormatter/index.tsx [hunk 5]
```diff
@@ -23,31 +24,42 @@ export interface AmountFormatterProps {
...
+    if (amount == null) return <>-</>
 
-  const amountString = formatBaseUnitsAsRose(props.amount, {
-    minimumFractionDigits: props.minimumFractionDigits ?? 1,
-    maximumFractionDigits: props.maximumFractionDigits ?? 15,
-  })
+    const formatter = baseUnit ? formatBaseUnitsAsRose : formatWeiAsWrose
+    const amountString = formatter(amount, {
+      minimumFractionDigits: minimumFractionDigits ?? 1,
+      maximumFractionDigits: maximumFractionDigits ?? baseUnit ? 15 : 18,
```
Comment (lukaw3d) [inline src/app/components/AmountFormatter/index.tsx:43]: This has wrong priority of operations

File: src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx [hunk 6]
```diff
@@ -23,20 +23,25 @@ export const TransactionConfirmation = () => {
               ? t('paraTimes.confirmation.receiving', 'into the receiving')
               : t('paraTimes.confirmation.withdrawing', 'from the withdrawing'),
             address: transactionForm.recipient,
-            isEvmc: isEvmcParaTime ? t('paraTimes.common.evmc', '(EVMc)') : '',
+            paratimeType: isEvmcParaTime ? t('paraTimes.common.evmcType', '(EVMc)') : '',
             paraTime: paraTimeName,
             value: transactionForm.amount,
           }}
-          defaults="You are about to transfer <strong>{{value}} ROSE</strong> tokens {{actionType}} wallet on the <strong>{{paraTime}}</strong> {{isEvmc}} ParaTime: <strong>{{address}}</strong>"
+          defaults="You are about to transfer <strong>{{value}} ROSE</strong> tokens {{actionType}} wallet on the <strong>{{paraTime}}</strong> {{paratimeType}} ParaTime: <strong>{{address}}</strong>"
```
Comment (lukaw3d) [inline src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx:30]: `<strong>{{value}} ROSE</strong>`

should use selectTicker, so it displays TEST on testnet

File: src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    <ParaTimeContent
+      description={
+        <Trans
+          i18nKey="paraTimes.confirmation.description"
+          t={t}
+          values={{
+            actionType: isDepositing
+              ? t('paraTimes.confirmation.receiving', 'into the receiving')
+              : t('paraTimes.confirmation.withdrawing', 'from the withdrawing'),
+            address: transactionForm.recipient,
```
Comment (lukaw3d) [inline src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx:25]: This is empty when withdrawing from EVM.

Unless you first go through withdrawing from cipher, click "Previous", and withdraw from Emerald: then it displays a cipher address :scream:

File: src/app/components/AmountFormatter/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukaw3d) [reply src/app/components/AmountFormatter/index.tsx:16]: I guess there's no good boolean name that describes base unit vs wei.

So: `amountUnit?: 'baseUnits' | 'wei'` or `unitOfAmount?: 'nanoRose' | 'attaRose'`

File: src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx (hunk 7 above)
Comment (buberdds) [reply src/app/pages/ParaTimesPage/TransactionConfirmation/index.tsx:25]: EMVc withdraws need an extra step with recipient address. will add it to TODO list and fix while working on withdraws.

File: src/app/state/paratimes/saga.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    const nativeDenominationHex = oasis.misc.toHex(token.NATIVE_DENOMINATION)
+    for (const [denomination, amount] of accountBalances.balances) {
+      const denominationHex = oasis.misc.toHex(denomination)
+      if (denominationHex === nativeDenominationHex) {
+        nativeDenominationBalanceBI = oasis.quantity.toBigInt(amount)
+        break
+      }
+    }
+  }
+  return nativeDenominationBalanceBI
```
[the remark this answers is not in the document]
Comment (buberdds) [reply src/app/state/paratimes/saga.ts]: I asked the same question a week ago and the answer was yes

File: src/app/lib/helpers.ts [hunk 9]
```diff
@@ -51,26 +51,50 @@ export function concat(...parts: Uint8Array[]) {
...
+  )
+}
+
+export function isAmountGreaterThan(amount: StringifiedBigInt, value: StringifiedBigInt) {
+  return parseRoseStringToBigNumber(amount).isGreaterThan(new BigNumber(value))
+}
+
+export function isEvmcAmountGreaterThan(amount: StringifiedBigInt, value: StringifiedBigInt) {
+  return parseRoseStringToBigNumber(amount, 18).isGreaterThan(new BigNumber(value))
+}
```
[the remark this answers is not in the document]
Comment (buberdds) [reply src/app/lib/helpers.ts:82]: moved to TODO

File: src/app/components/AmountFormatter/index.tsx [hunk 10]
```diff
@@ -23,31 +24,43 @@ export interface AmountFormatterProps {
...
+    if (amount == null) return <>-</>
 
-  const amountString = formatBaseUnitsAsRose(props.amount, {
-    minimumFractionDigits: props.minimumFractionDigits ?? 1,
-    maximumFractionDigits: props.maximumFractionDigits ?? 15,
-  })
+    const formatter = isUsingBaseUnits ? formatBaseUnitsAsRose : formatWeiAsWrose
+    const amountString = formatter(amount, {
+      minimumFractionDigits: minimumFractionDigits ?? 1,
+      maximumFractionDigits: maximumFractionDigits ? maximumFractionDigits : isUsingBaseUnits ? 15 : 18,
```
Comment (lukaw3d) [inline src/app/components/AmountFormatter/index.tsx:44]: I don't think this works well either, we use maximumFractionDigits={0} in some places :D

File: src/app/pages/ParaTimesPage/TransactionAmount/index.tsx [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+          i18nKey="paraTimes.amount.description"
+          t={t}
+          values={{
+            actionType: isDepositing
+              ? t('paraTimes.amount.receiving', 'to the receiving')
+              : t('paraTimes.amount.withdrawing', 'from the withdrawing'),
+            paratimeType: isEvmcParaTime ? t('paraTimes.common.evmcType', '(EVMc)') : '',
+            paraTime: paraTimeName,
+          }}
+          defaults='Please enter the amount of ROSE tokens you wish to transfer {{actionType}} wallet on the <strong>{{paraTime}}</strong> {{paratimeType}} ParaTime and then click "Next"'
```
Comment (lukaw3d) [inline src/app/pages/ParaTimesPage/TransactionAmount/index.tsx:78]: Another ROSE => {{ticker}}

File: src/app/pages/ParaTimesPage/ParaTimeSelection/index.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    <ParaTimeContent
+      description={
+        isDepositing
+          ? t(
+              'paraTimes.selection.depositDescription',
+              'Please select which ParaTime you wish to transfer your ROSE tokens to and then click "Next".',
+            )
+          : t(
+              'paraTimes.selection.withdrawDescription',
+              'Please select which ParaTime you wish to withdraw your ROSE tokens from and then click "Next".',
```
Comment (lukaw3d) [inline src/app/pages/ParaTimesPage/ParaTimeSelection/index.tsx:49]: ROSE => {{ticker}}

**Review by lukaw3d — APPROVED** [review]

**Pull request merged**